### PR TITLE
Build journal index on demand

### LINKED ISSUE #12667: Reduce startup time when there are many segments
**Description**

Currently, on startup, we scan all segments to detect corruption and to re-build journal index. This result in a high startup duration.

For example, in the following experiment, a broker took almost 2 minutes to startup when there were around 60 segments per partition.
image

A partition can end up with large number of segments usually if the exporting is slow or not making progress. Other unexpected bugs in processing or snapshot can also lead to the situation. Although it is important to fix the root cause, we should also try to reduce the impact of it on startup duration. We have observed that high startup duration make things worse, because during that time the broker is not ready, it may not receive events until it has scanned all segments, and as a result needs longer time to catch up with the leader. If more than one broker is restarted, the cluster is unavailable for a long time. This has resulted in a few incidents in production. 

To improve this situation, we should remove the scan of all segments during startup. This has the following impact.
1. We cannot detect corruption on startup. But this might be ok, because we will detect it later when the corrupted entry is read. If it is never read, then it would be ok to ignore it. An entry is never read again if it was already replicated to all brokers, exported, and processed, which means it will be compacted soon.
2. We cannot rebuilt the journal index, which means the seeks might be slow. We should verify how much impact it has, and see if we can improve it. Some ideas - a. persist the index b. dynamically build it when we think it is needed. c. build it async in background.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

With PR #12839 , we do not built index of segment entries for existing segments after restart. This can impact seek time, as the seek has to read all entries to reach the desired index. In this PR, we built the index on demand - that is the  first time when a reader seek to an unindexed entry, it index the entries that it read while seeking. If the reader has to seek again, then it can use the index. There are cases where reader has to seek twice such in seekToAsqn, or in rare cases, if the follower fails frequently, leader has to reset the index by seeking. 

Pros:
- Only index if necessary. This operation is cheap as the seek has to read the entries anyway and indexing is cheap. 

Cons:
- First seek can still be slow if the entry is not at the beginning of a segment.
- The index re-built may not be used again, if there is not another seek in the same segment.

Alternative approach:
- Built index in a background thread after startup. However, I'm not sure if the complexity is worth it as the segments will be most likely deleted soon.

## Related issues

closes #12667

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by deepthidevaki** [comment]
@npepinpe This is followup of PR that removes building index on startup. I tried to keep it simple with the solution. Happy to discuss if you think we should handle it differently.

File: journal/src/main/java/io/camunda/zeebe/journal/file/JournalMetrics.java [hunk 1]
```diff
@@ -102,6 +103,15 @@ final class JournalMetrics {
           .labelNames(PARTITION_LABEL)
           .register();
 
+  private static final Histogram SEEK_LATENCY =
```
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/JournalMetrics.java:106]: :heart:

File: journal/src/main/java/io/camunda/zeebe/journal/file/SparseJournalIndex.java [hunk 2]
```diff
@@ -18,12 +18,15 @@
...
 import java.util.TreeMap;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
 
 final class SparseJournalIndex implements JournalIndex {
 
   private final int density;
-  private final TreeMap<Long, Integer> indexToPosition = new TreeMap<>();
-  private final TreeMap<Long, Long> asqnToIndex = new TreeMap<>();
+  private final ConcurrentNavigableMap<Long, Integer> indexToPosition =
```
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SparseJournalIndex.java:27]: :wrench: As we only update some things from multiple threads, maybe we want to annotate/document that (i.e. why is `indexToAsqn` not thread-safe). I don't know how likely this is to cause confusion in the future...though it took me a good 10 seconds to understand why this change, but maybe that's just me :smile:

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournalReader.java [hunk 3]
```diff
@@ -109,46 +117,49 @@ public long seekToAsqn(final long asqn) {
...
-      }
+    try (final var ignored = metrics.observeSeekLatency()) {
+      final var stamp = journal.acquireReadlock();
+      try {
+        final var journalIndex = journal.getJournalIndex();
+        final var index = journalIndex.lookupAsqn(asqn, indexUpperBound);
+
+        // depending on the type of index, it's possible there is no ASQN indexed, in which case
+        // start
+        // from the beginning
```
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournalReader.java:128]:
```suggestion
        // start from the beginning
```

**Review by npepinpe — APPROVED** [review]
:rocket: 

:x: There are no tests to check that we're indexing when reading, so no real test for this feature. Just testing the happy path would be enough imo.

No need for another review :)

File: journal/src/main/java/io/camunda/zeebe/journal/file/SparseJournalIndex.java (hunk 2 above)
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/SparseJournalIndex.java:27]: Actually, I think `indexToAsqn` should also be thread safe. Now, the readers while seeking can update the it. :face_with_head_bandage:

**Comment by deepthidevaki** [comment]
bors merge

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**